### PR TITLE
feat: adding kapp package

### PR DESCRIPTION
**Pull request by renuy**
Signed-off-by: Renu Yarday <renuyarday@gmail.com>

### CODE REVIEW

**Comment by renuy** [comment]
cc: @cppforlife

**Comment by gerhard** [comment]
It's nice to see you here @cppforlife @renuy 👋🏻

File: pkg/universe.dagger.io/kapp/cue.mod/pkg/github.com/renuy/kapp/source.cue
Comment (TomChv) [inline pkg/universe.dagger.io/kapp/cue.mod/pkg/github.com/renuy/kapp/source.cue:11]: Should be named `#CLI`.

To avoid `kapp.#Kapp`, instead `kapp.#CLI` is more natural.

**Review by TomChv — DISMISSED** [review]
Experimental package should be in `pkg/universe.dagger.io/x/`, can you move it in?

As well, your package is for pre europa version (v0.1). 
Dagger is now v0.2 and a lot of things have change in the API.

Could you migrate your package to europa so we can merge it?

To help you, here's some example of package from the new universe that will help you to migrate
- [Pull, build, run container with `docker` package](https://github.com/dagger/dagger/tree/main/pkg/universe.dagger.io/docker)
- [Run script with bash.#Run](https://github.com/dagger/dagger/blob/main/pkg/universe.dagger.io/bash/bash.cue#L14)

Indeed, some documentation about Dagger 0.2:
- [Official doc](https://docs.dagger.io)
- [Europa Core concept](https://docs.dagger.io/category/core-concepts)
- [Core action reference](https://docs.dagger.io/1222/core-actions-reference)
- [Interact with host's filesystem](https://docs.dagger.io/1203/client)

And a package made by one of our contributor
-  [Pulumi package](https://github.com/dagger/dagger/blob/main/pkg/universe.dagger.io/x/david%40rawkode.dev/pulumi/pulumi.cue)
- [Pulumi PR](https://github.com/dagger/dagger/pull/1991)

**Comment by renuy** [comment]
@TomChv , difficult to find documentation pertaining to v0.2 (have asked on discord too)
could you let me know what are the corresponding commands for (To test):
$ dagger init
$ dagger new kapp-test -p ./test
$ dagger input text TestKubeconfig -f "$HOME"/.kube/config -e kapp-test
$ dagger up -e kapp-test

**Comment by TomChv** [comment]
> $ dagger init
$ dagger new kapp-test -p ./test

Now it's called `dagger project init` and you can do `dagger project update` to install dagger universe.

> $ dagger input text TestKubeconfig -f "$HOME"/.kube/config -e kapp-test

No more inputs, no you will need to use `client` to retrieve file from host.

> $ dagger up -e kapp-test

There is no more `dagger up`, it's `dagger do` now

But to test, you'll need to change the whole configuration, there is no more `#up` too, I encourage you to read links from my previous command to understand how Europa 2.0 works.

**Comment by renuy** [comment]
@shykes @TomChv Doing second pass with feature. do take a look.
Thanks!

**Comment by renuy** [comment]
To run tests
cd dagger/pkg/universe.dagger.io/x/carvel.dev/test
dagger do --help
dagger do test

**Comment by helderco** [comment]
Hey, thank you for your contribution.

Allow me a quick reminder to point you to these docs:
- [Package Coding Style](https://docs.dagger.io/1226/coding-style)
- [Contributing to Dagger](https://docs.dagger.io/1227/contributing)

You also need a _bats_ file for the tests.

**Comment by renuy** [comment]
> Hey, thank you for your contribution.
> 
> Allow me a quick reminder to point you to these docs:
> 
> * [Package Coding Style](https://docs.dagger.io/1226/coding-style)
> * [Contributing to Dagger](https://docs.dagger.io/1227/contributing)
> 
> You also need a _bats_ file for the tests.

Done

File: pkg/universe.dagger.io/x/carvel.dev/kapp.cue
Comment (TomChv) [inline pkg/universe.dagger.io/x/carvel.dev/kapp.cue:195]: Nit: `docker.#Build` is useless since you have a single step

File: pkg/universe.dagger.io/x/carvel.dev/test/test.bats
Comment (TomChv) [inline pkg/universe.dagger.io/x/carvel.dev/test/test.bats:11]: Since we do not deploy a kubernetes cluster in the CI, those test will never succeed.
@aluzzardi Should we deploy a kind cluster like we did pre europa?

**Review by TomChv — COMMENTED** [review]
Thank for your refactor :rocket:

There are some improvement possible to remove code duplication but we can merge the package like that and take care of those later.

Make sure that CI pass and we could merge that one.

File: pkg/universe.dagger.io/x/carvel.dev/kapp.cue
Comment (renuy) [reply pkg/universe.dagger.io/x/carvel.dev/kapp.cue]: Done

File: pkg/universe.dagger.io/x/carvel.dev/kapp.cue
Comment (renuy) [reply pkg/universe.dagger.io/x/carvel.dev/kapp.cue]: Done

File: pkg/universe.dagger.io/x/carvel.dev/test/test.bats
Comment (renuy) [reply pkg/universe.dagger.io/x/carvel.dev/test/test.bats]: do let me know, what is needed here. Should I remove test.bats?

Comment (TomChv) [reply pkg/universe.dagger.io/x/carvel.dev/test/test.bats]: I think you can comment it, that should be enough for now

**Review by shykes — APPROVED** [review]
Proactively approving this, since it's an experimental package and does not need to be fully functional to be merged.

@renuy we'll wait for you to tell us when you'd like us to merge - feel free to make further changes either before or after merging. Code reviews are here as a convenience to help you improve your package.

**Comment by jpadams** [comment]
> Experimental package should be in `pkg/universe.dagger.io/x/`, can you move it in?
> 
> As well, your package is for pre europa version (v0.1). Dagger is now v0.2 and a lot of things have change in the API.
> 
> Could you migrate your package to europa so we can merge it?
> 
> To help you, here's some example of package from the new universe that will help you to migrate
> 
> * [Pull, build, run container with `docker` package](https://github.com/dagger/dagger/tree/main/pkg/universe.dagger.io/docker)
> * [Run script with bash.#Run](https://github.com/dagger/dagger/blob/main/pkg/universe.dagger.io/bash/bash.cue#L14)
> 
> Indeed, some documentation about Dagger 0.2:
> 
> * [Official doc](https://docs.dagger.io)
> * [Europa Core concept](https://docs.dagger.io/category/core-concepts)
> * [Core action reference](https://docs.dagger.io/1222/core-actions-reference)
> * [Interact with host's filesystem](https://docs.dagger.io/1203/client)
> 
> And a package made by one of our contributor
> 
> * [Pulumi package](https://github.com/dagger/dagger/blob/main/pkg/universe.dagger.io/x/david%40rawkode.dev/pulumi/pulumi.cue)
> * [Pulumi PR](https://github.com/dagger/dagger/pull/1991)

This work is now done. I'm clearing the "Changes Requested" blocker.

**Review by jpadams — APPROVED** [review]
Thanks for contrib!

**Comment by jpadams** [comment]
To test on my Mac, I added `kapp` via homebrew: https://carvel.dev/kapp/docs/v0.46.0/install/#via-homebrew-macos-or-linux

and turned on kubernetes in Docker Desktop. That cleared my first test error with `kapp ls`.

Working on error with `kapp inspect`.

**Comment by jpadams** [comment]
By commenting various parts of the tests, I see there is non-deterministic ordering of the test "sub-actions". Sometimes the app gets created in time to inspect it or delete it, sometimes not. Likely need to add deps between steps.

File: pkg/universe.dagger.io/x/carvel.dev/test/test.bats
Comment (renuy) [reply pkg/universe.dagger.io/x/carvel.dev/test/test.bats]: deleted the file for now. was unable to run it locally, so havent added it.

**Comment by renuy** [comment]
@jpadams , ready to merge

File: pkg/universe.dagger.io/x/carvel.dev/kapp/kapp.cue
Comment (cppforlife) [inline pkg/universe.dagger.io/x/carvel.dev/kapp/kapp.cue:9]: how come it only works with kind?

File: pkg/universe.dagger.io/x/carvel.dev/kapp/kapp.cue
Comment (cppforlife) [inline pkg/universe.dagger.io/x/carvel.dev/kapp/kapp.cue:27]: any way to do this without a shell?

File: pkg/universe.dagger.io/x/carvel.dev/kapp/kapp.cue
Comment (cppforlife) [inline pkg/universe.dagger.io/x/carvel.dev/kapp/kapp.cue:68]: nit/style: space after `//`

File: pkg/universe.dagger.io/x/carvel.dev/kapp/kapp.cue
Comment (cppforlife) [inline pkg/universe.dagger.io/x/carvel.dev/kapp/kapp.cue:21]: instead of doing this conditional logic in bash, why dont we do it in cue when constructing command. (goes with the below comment of how we can get rid of using a shell)

File: pkg/universe.dagger.io/x/carvel.dev/kapp/kapp.cue
Comment (jpadams) [reply pkg/universe.dagger.io/x/carvel.dev/kapp/kapp.cue:106]: @cppforlife I'm going to merge and then let's follow up with PRs for more improvements :) Thank you for the thoughtful review.

**Pull request merged**
